cpp/util/ManagedArray.h

```diff
@@ -0,0 +1,146 @@
+ *  and any such array can resize or reallocate this data. The pointer to
+ *  pointer infrastructure ensures that such changes properly propagate to all
+ *  ManagedArrays referencing a given memory space. In addition, this
+ *  infrastructure allows the creation of a completely new ManagedArray with a
+ *  new set of pointers that also manages the same data, allowing it to keep
+ *  the original array alive if the original ManagedArray instances become
+ *  decoupled from it.
+ */
+template<typename T> class ManagedArray
+{
+public:
+    //! Default constructor with optional size for thread local arrays
+    /*! \param size Size of the array to allocate.
+     */
+    ManagedArray(unsigned int size=0)
+    {
+        m_size = std::make_shared<unsigned int>(size);
+        // Hard code making an array of at least size 1.
+        m_data = std::shared_ptr<std::shared_ptr<T> >(
+            new std::shared_ptr<T>(new T[1], std::default_delete<T[]>()));
+        create_new_array(size);
+        reset();
+    }
+
+    //! Destructor (currently empty because data is managed by shared pointer).
+    ~ManagedArray() {}
+
+    //! Update size of the array.
+    /*! \param size New size of the array.
+     */
+    void resize(unsigned int size)
+    {
+        if (size != *m_size)
+        {
+            *m_size = size;
+            reallocate();
+        }
+    }
+
+    //! Allocate new memory for the array.
+    /*! This function primarily serves as a way to reallocate new memory from
+     *  the Python API if needed.
+     * \param size New size of the array.
+     */
+    void reallocate()
+    {
+        create_new_array(*m_size);
+        reset();
+    }
+
+    //! Reset the contents of array to be 0.
+    void reset()
+    {
+        if (*m_size != 0)
+        {
+            memset((void*) get(), 0, sizeof(T) * (*m_size));
+        }
+    }
+
+    //! Return the underlying pointer (requires two levels of indirection).
+    T *get() const
+    {
+        std::shared_ptr<T> * tmp = m_data.get();
+        return (*tmp).get();
+    }
+
+    //! Writeable index into array.
+    T &operator[](unsigned int index)
+    {
+        if (index >= *m_size)
+        {
+            throw std::runtime_error("Attempted to access data out of bounds.");
+        }
+        return get()[index];
+    }
+
+    //! Read-only index into array.
+    const T &operator[](unsigned int index) const
+    {
+        if (index >= *m_size)
+        {
+            throw std::runtime_error("Attempted to access data out of bounds.");
+        }
+        return get()[index];
+    }
+
+    //! Get the size of the current array.
+    unsigned int size() const
+    {
+        return *m_size;
+    }
+
+    //! Make a deep copy of this array (i.e. copy the underlying data).
+    /*! This function returns by value since all memory ownership information
+     *  is transmitted via shared pointers. 
+     */
+    void *deepCopy()
```

Comment:
Is there any particular reason this is a void pointer instead of `ManagedArray<T>*`?

Reply:
No. At one point the output of `deepCopy` was getting passed through additional function calls on the Cython side and I was just being lazy and saving myself a cast from `ManagedArray<T> *` to `void *` in Cython. Definitely no reason to leave this long term.